From 613d4f89630488d521fbf7f6dfd20296cfaac8f5 Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Sun, 28 May 2023 23:07:42 +0200
Subject: [PATCH] [Fix] Fix payment requests certificates and
 logos_feature_minerfund.py test (#300)

    * [backport] [paymentservertests] fix certificates and payment requests

    Summary:
    Procedure:

    ```
    git clone git@github.com:gavinandresen/paymentrequest.git
    cd paymentrequest
    ca_in_a_box/create_ca.sh
    ca_in_a_box/create_intermediate_cas.sh

    openssl x509 -in ca_in_a_box/certs/cacert.pem -inform PEM -outform DER | openssl enc -base64

    cd c++
    make

    mv ../ca_in_a_box .

    ./paymentrequest-create paytoaddress=1Nf4eZ76SFZdhrHVNLQkLAiS1oPPGaZxSv  amount=42 expires=1986627327 | ./paymentrequest-dump
    ./paymentrequest-create paytoaddress=1Nf4eZ76SFZdhrHVNLQkLAiS1oPPGaZxSv  amount=42 expires=1986627327 | base64

    ./paymentrequest-create paytoaddress=1Nf4eZ76SFZdhrHVNLQkLAiS1oPPGaZxSv  amount=42 expires=4826525372 payment_url=testmerchant8.org certificates=./ca_in_a_box/intermediate_8/certs/demomerchant.pem,./ca_in_a_box/intermediate_8/certs/cacert.pem,./ca_in_a_box/intermediate_7/certs/cacert.pem,./ca_in_a_box/intermediate_6/certs/cacert.pem,./ca_in_a_box/intermediate_5/certs/cacert.pem,./ca_in_a_box/intermediate_4/certs/cacert.pem,./ca_in_a_box/intermediate_3/certs/cacert.pem,./ca_in_a_box/intermediate_2/certs/cacert.pem,./ca_in_a_box/intermediate_1/certs/cacert.pem,./ca_in_a_box/certs/cacert.pem privatekey=./ca_in_a_box/intermediate_8/private/demomerchantkey.pem |  base64
    ```

    The patch to fix the outdated OpenSSL code is:
    ```
    diff --git a/c++/paymentrequest-create.cpp b/c++/paymentrequest-create.cpp
    index 5fdbc8c..69d5b8d 100644
    --- a/c++/paymentrequest-create.cpp
    +++ b/c++/paymentrequest-create.cpp
    @@ -178,16 +178,17 @@ bool decode_base58(const string& btcaddress, unsigned char& version, string& has
         unsigned char decoded[25];

         size_t nBytes = 0;
    -    BIGNUM bn58, bn, bnChar;
    +    BIGNUM *bn58, *bn, *bnChar;
         BN_CTX *ctx;

         ctx = BN_CTX_new();
    -    BN_init(&bn58);
    -    BN_init(&bn);
    -    BN_init(&bnChar);

    -    BN_set_word(&bn58, 58);
    -    BN_set_word(&bn, 0);
    +    bn58 = BN_new();
    +    bn = BN_new();
    +    bnChar = BN_new();
    +
    +    BN_set_word(bn58, 58);
    +    BN_set_word(bn, 0);

         for (unsigned int i = 0; i < btcaddress.length(); i++) {
             const char *p1 = strchr(base58_chars, btcaddress[i]);
    @@ -195,23 +196,23 @@ bool decode_base58(const string& btcaddress, unsigned char& version, string& has
                 goto out;
             }

    -        BN_set_word(&bnChar, p1 - base58_chars);
    +        BN_set_word(bnChar, p1 - base58_chars);

    -        assert(BN_mul(&bn, &bn, &bn58, ctx));
    -        assert(BN_add(&bn, &bn, &bnChar));
    +        assert(BN_mul(bn, bn, bn58, ctx));
    +        assert(BN_add(bn, bn, bnChar));
         }

    -    nBytes = BN_num_bytes(&bn);
    +    nBytes = BN_num_bytes(bn);
         if (nBytes == 0 || nBytes > 25)
             return false;

         std::fill(decoded, decoded+25, (unsigned char)0);
    -    BN_bn2bin(&bn, &decoded[25-nBytes]);
    +    BN_bn2bin(bn, &decoded[25-nBytes]);

     out:
    -    BN_clear_free(&bn58);
    -    BN_clear_free(&bn);
    -    BN_clear_free(&bnChar);
    +    BN_clear_free(bn58);
    +    BN_clear_free(bn);
    +    BN_clear_free(bnChar);
         BN_CTX_free(ctx);

         version = decoded[0];
    diff --git a/c++/paymentrequest-dump.cpp b/c++/paymentrequest-dump.cpp
    index d20e400..d28d1dd 100644
    --- a/c++/paymentrequest-dump.cpp
    +++ b/c++/paymentrequest-dump.cpp
    @@ -133,12 +133,14 @@ int main(int argc, char **argv) {
         request.SerializeToString(&data_to_verify);
         request.set_signature(signature);

    -    EVP_MD_CTX ctx;
    +    EVP_MD_CTX *ctx;
    +    ctx = EVP_MD_CTX_new();
    +
         EVP_PKEY *pubkey = X509_get_pubkey(signing_cert);
    -    EVP_MD_CTX_init(&ctx);
    -    if (!EVP_VerifyInit_ex(&ctx, digestAlgorithm, NULL) ||
    -        !EVP_VerifyUpdate(&ctx, data_to_verify.data(), data_to_verify.size()) ||
    -        !EVP_VerifyFinal(&ctx, (const unsigned char*)signature.data(), signature.size(), pubkey)) {
    +    EVP_MD_CTX_init(ctx);
    +    if (!EVP_VerifyInit_ex(ctx, digestAlgorithm, NULL) ||
    +        !EVP_VerifyUpdate(ctx, data_to_verify.data(), data_to_verify.size()) ||
    +        !EVP_VerifyFinal(ctx, (const unsigned char*)signature.data(), signature.size(), pubkey)) {

             printf("Bad signature, invalid PaymentRequest.\n");
         }

    ```

    Notes:
     - the merchant certificate is the first in the chain, then all the intermediate certificate authorities in reverse order, and finally the root certificate authority. The private key is the one for the merchant.
     - do not run the `create_ca.sh` and `create_intermediate_cas.sh` more than once, or delete the previous results in between two runs. This is because re-running it causes some or all keys to be overwritten, but it fails to recreate the new corresponding certs.

    Test Plan: `ninja check-bitcoin-qt`

    Reviewers: #bitcoin_abc, Fabien

    Reviewed By: #bitcoin_abc, Fabien

    Differential Revision: https://reviews.bitcoinabc.org/D12887

    * [Fix] Fix logos_feature_minerfund after the upgrade

    This test has been broken after the December upgrade, and this fixes it.

Metadata recovery: the historical body is quoted above because it embeds
a diff; importing it verbatim would treat that example as source changes.

Rebase provenance: this historical change was adapted onto the pinned
Bitcoin ABC base. Source diff is the recorded Lotus V1 replay commit below.

Historical-source-commit: 581a3ae422350f1bba70be21c4d603b5810b43fd
Recorded-replay-commit: 3da2c7bea74cb4e56dbd30a7470c65c89773c992

diff --git a/test/functional/logos_feature_minerfund.py b/test/functional/logos_feature_minerfund.py
index e3d060bf5a..e208e007ab 100755
--- a/test/functional/logos_feature_minerfund.py
+++ b/test/functional/logos_feature_minerfund.py
@@ -73,25 +73,14 @@ class MinerFundTest(BitcoinTestFramework):
         prepare_block(block)
         assert_equal(node.submitblock(ToHex(block)), 'bad-cb-minerfund')
 
-        # Submit a custom block that does send the required outputs
-        # with the miner reward output in the middle (which is allowed)
-        # but with the outputs in the wrong order (which is not allowed)
+        # Submit a custom block that sends the required outputs
+        # with the miner reward at the end (which is allowed)
         coinbase.vout[1:] = []
         for output in expected_outputs:
             coinbase.vout.append(CTxOut(int(output['value']),
                                         CScript(bytes.fromhex(output['scriptPubKey']))))
         coinbase.vout.insert(2, CTxOut(int(SUBSIDY / 2 * COIN),
                                        CScript([OP_HASH160, bytes(20), OP_EQUAL])))
-        coinbase.vout[2], coinbase.vout[4] = coinbase.vout[4], coinbase.vout[2]
-        coinbase.rehash()
-        block = create_block(int(best_block_hash, 16),
-                             coinbase, block_time + 1)
-        block.nHeight = block_height
-        prepare_block(block)
-        assert_equal(node.submitblock(ToHex(block)), 'bad-cb-minerfund')
-
-        # Fix the order and block is valid
-        coinbase.vout[2], coinbase.vout[4] = coinbase.vout[4], coinbase.vout[2]
         coinbase.rehash()
         block = create_block(int(best_block_hash, 16),
                              coinbase, block_time + 1)
@@ -116,6 +105,8 @@ class MinerFundTest(BitcoinTestFramework):
         coinbase = create_coinbase(block_height)
         coinbase.vout[1] = CTxOut(int(share_multiplier * SUBSIDY / 2 * COIN),
                                   CScript([OP_HASH160, bytes(20), OP_EQUAL]))
+        block_template = node.getblocktemplate()
+        expected_outputs = block_template['coinbasetxn']['minerfund']['outputs']
         for output in expected_outputs:
             coinbase.vout.append(CTxOut(int(share_multiplier * output['value']),
                                         CScript(bytes.fromhex(output['scriptPubKey']))))
